### PR TITLE
Upgrades to browse-everything 2.x-stable

### LINKED ISSUE #3505: Bulk ingest of PDFs not pulling source metadata
When bulk ingesting PDFs it's not pulling the metadata identifier from the folder name, and instead just naming it after the file that's being ingested.

To test this is fixed:

Ingest a folder of PDFs from https://figgy.princeton.edu/bulk_ingest/scanned_resource in the structure
```
- Parent Folder (selected)
  - 123456
    - file.pdf
  - 802310
    - file2.pdf
```

Ensure that the ScannedResources it creates are set with the SourceMetadataIdentifier `123456` and `802310`

### LINKED ISSUE #3591: File upload should work if I selected the file by clicking the name
when im trying to select a file to submit for file upload from pudl share to figgy: if i click on the name versus checking the checkbox directly, the system freezes at 10% complete and wont allow me to click submit as an option.

reported by @kelea99 

Screen Shot 2020-01-16 at 1 06 36 PM (3)

### LINKED ISSUE #3604: Bulk ingest for single items are ingested as MVWs
As reported by a user:
> Lapidus ingest went in as a title-only MVW instead of 49 unique title-only items.

Please see https://figgy.princeton.edu/catalog/8f43d88e-cb3e-45ab-bf2d-e5f28c6089e5 for the failed ingest.

The files can be found on the isilon mount in `pudl/hudson_lapidus/ingest`

### LINKED ISSUE #3606: Browse Everything files come through twice
[Figgy/production] NoMethodError: undefined method `split' for nil:NilClass

## Backtrace

[View full backtrace and more info at honeybadger.io](https://app.honeybadger.io/projects/53391/faults/60157535)

### LINKED ISSUE #3724: Test integrating the BrowseEverything 2.0.0.pre.alpha1 release
This work follows what I have been introducing on https://github.com/samvera/browse-everything/compare/2.x-stable...jrgriffiniii:issues-319-jrgriffiniii-google-picker and https://github.com/samvera-labs/browse-everything-redux-react/compare/master...jrgriffiniii:issues-319-jrgriffiniii-google-filepicker.

I currently have Google Drive supported using the Google Picker API with these branches, and I've tested recursive uploads using the Rails Engine and React interfaces in isolation. Once I've successfully tested integration with Figgy, I will work with the other contributors to release an initial `alpha` version (given that the drivers for Dropbox, Amazon S3, and Box will not be supported, using `alpha` seems justified).

### LINKED ISSUE #3791: Bulk Ingest not working
@roelmoon attempt to ingest Lapidus items failed this week.

### PR DESCRIPTION
**Pull request by jrgriffiniii**
This will upgrade to the first browse-everything 2.0.0 alpha pre-release. This shall include integrated support for the the React user interface (which also integrates the Google Picker widget for supporting Google Drive uploads).

This is blocked by https://github.com/pulibrary/figgy/pull/3752 due to its dependency on NodeJS 12.x releases.

fixes #3604 
fixes #3791 
fixes #3724 
fixes #3591 
fixes #3606
fixes #3505 

On merge please resolve https://app.honeybadger.io/projects/53391/faults/60157535

TODO: 

upstream:

- [x] namespace the classes in https://github.com/samvera/browse-everything/blob/2.x-stable/lib/browse_everything/authorization.rb#L10-L18
- [x] Remove `mint_new_id` from https://github.com/samvera/browse-everything/blob/2.x-stable/app/jobs/browse_everything/upload_job.rb#L61

locally:
- [x] Upgrade to rails 5.2 
- [x] Rename BrowseEverythingIngester
- [x] Fix staging database (@hackmastera )
- [x] Use shared drive for activestorage files  (@hackmastera )
- [x] merge #3834 (don't copy local files, clean up tmp dir)
- [x] CloudIngester should happen in the background. (@tpendragon)
- [x] QA all the tickets we think this fixes
- [x] remove MVW option from page (@tpendragon)
- [x] document required directory structure on page
- [x] Test for the source metadata identifier behavior (should work for finding aids components/bibids) (@tpendragon)
- [x] fix test coverage
- [x] Make `spec/resources/scanned_resources/scanned_resources_controller_spec.rb` not stub `BrowseEverything::UploadJob`, and send real files.
- [x] Adjust cloud ingest specs in `spec/controllers/bulk_ingest_controller_spec.rb` to have accurate `BrowseEverything::Uploads`, do not stub `BrowseEverything::UploadJob`, and instead stub the Google Drive provider's `#find_bytestream` and `#find_container` methods.
- [x] Test that BE is cleaning up files?
- [x] Rebase this PR
- [x] try deleting that new block from file_appender (ensure pending_uploads on all resources are blank post-ingest)

### CODE REVIEW

**Comment by jrgriffiniii** [comment]
The header for the layout seems to no longer be rendering in the staging environment, but does so for me locally:

image

image

**Comment by jrgriffiniii** [comment]
https://circleci.com/gh/pulibrary/figgy/25098#tests/containers/1 is failing due to tests which I cannot trigger failures for locally:

```
Failures:

  1) ArchivalMediaCollectionDecorator raw imported metadata is not displayed
     Failure/Error: h.visibility_badge(visibility, public_readable_state?)
     
     NoMethodError:
       undefined method `visibility_badge' for #<#<Class:0x0000563ebc0646a8>:0x0000563ebca191a8>
     # ./app/decorators/valkyrie/resource_decorator.rb:48:in `block in visibility'
     # ./app/decorators/valkyrie/resource_decorator.rb:47:in `map'
     # ./app/decorators/valkyrie/resource_decorator.rb:47:in `visibility'
     # ./app/decorators/application_decorator.rb:39:in `[]'
     # ./app/decorators/application_decorator.rb:63:in `block in to_h'
     # ./app/decorators/application_decorator.rb:62:in `map'
     # ./app/decorators/application_decorator.rb:62:in `to_h'
     # ./app/decorators/application_decorator.rb:45:in `display_attributes'
     # ./app/resources/collections/archival_media_collection_decorator.rb:13:in `display_attributes'
     # ./spec/resources/archival_media_collections/archival_media_collection_decorator_spec.rb:49:in `block (3 levels) in <top (required)>'

  2) ArchivalMediaCollectionDecorator#identifier displays the identifier
     Failure/Error: h.visibility_badge(visibility, public_readable_state?)
     
     NoMethodError:
       undefined method `visibility_badge' for #<#<Class:0x0000563ebc0646a8>:0x0000563ebca191a8>
```

**Comment by jrgriffiniii** [comment]
I've updated this in order to ensure that the BrowseEverything persistence models are properly namespaced, but pending the CircleCI build passing, this should prepare for the upstream release of the Gem.

**Comment by jrgriffiniii** [comment]
Unfortunately this cannot use a Gem pre-release until there is another Hyrax 2.x release (please see https://github.com/samvera/browse-everything/pull/335#issuecomment-598227508).

**Comment by jrgriffiniii** [comment]
This is breaking the Order Manager on the staging environment.

**Comment by jrgriffiniii** [comment]
The following error is logged when JavaScript dependencies is loaded in the staging environment:
```
clipboard.js:660 Uncaught TypeError: o is not a function
    at Module.<anonymous> (clipboard.js:660)
    at n (clipboard.js:66)
    at clipboard.js:260
    at o (clipboard.js:14)
    at Object.<anonymous> (clipboard.js:10)
    at Object.<anonymous> (application-ef8e1694b54505ac5770.js:2)
    at n (bootstrap:19)
    at Module.<anonymous> (auth_link_clipboard.js:1)
    at n (bootstrap:19)
    at Module.<anonymous> (application-ef8e1694b54505ac5770.js:2)
```

Updating `clipboard` with https://github.com/pulibrary/figgy/pull/3754/commits/f6a2059a3b442091dd55e1b840d7b415a412e6f3 did not resolve this.

**Comment by jrgriffiniii** [comment]
I've been trying to isolate the JavaScript packages which might be triggering this. Thus far, I have limited these changes to require only the following:

```
"@babel/preset-react": "^7.8.3",
"react": "^16.12.0",
"react-dom": "^16.12.0",
"react-redux": "^7.1.3",
"redux-logger": "^3.0.6",
"babel-loader": "^8.0.6",
"webpack-cli": "^3.3.11",
```

`redux-logger` and `webpack-cli` are required for running a local development environment. `@babel/preset-react` and `babel-loader` are necessary for integrating React into the Rails application, and the remaining dependencies are just for React itself.

Once the staging environment is clear, I will attempt walking back through additional dependencies in order to determine which of this is triggering the error with `clipboard`.

**Comment by jrgriffiniii** [comment]
Reducing these dependencies did not resolve the error.

**Comment by jrgriffiniii** [comment]
Simply disabling the `Clipboard` constructor with https://github.com/pulibrary/figgy/pull/3754/commits/55f3d37fbb959e4b7d47990184d3c541607502d7 solved the error, so I suspect that this is somehow related to the syntax of the constructor itself.

**Comment by jrgriffiniii** [comment]
The minified release of `Clipboard` does not break for my local environment, but it does on staging. When I use a CDN for the source (e. g. `https://cdn.jsdelivr.net/npm/clipboard@2.0.6/dist/clipboard.js`) it works.

**Comment by jrgriffiniii** [comment]
Adding `<script src="//cdn.jsdelivr.net/npm/clipboard@2.0.6/dist/clipboard.js"></script>` to the `app/views/layouts/application.html.erb` view partial seems to be the only way I can find to resolve this. Explicitly importing from the source using `import ClipboardJS from '../../../node_modules/clipboard/dist/clipboard.js'` simply raises the same error (I suspect that this somehow relates to a Webpack configuration setting used to transpile the ES6 from the `clipboard` dependency).

**Comment by jrgriffiniii** [comment]
This should be ready for review and testing.

**Comment by jrgriffiniii** [comment]
`EventGenerator::OrangelightEventGenerator` test suites may also be failing intermittently (please see https://circleci.com/gh/pulibrary/figgy/25247#tests/containers/3)

**Comment by jrgriffiniii** [comment]
I'm struggling to reproduce these failures by testing just these suites:

```bash
% bunx rspec --bisect spec/services/event_generator/geoblacklight_event_generator_spec.rb spec/services/event_generator/orangelight_event_generator_spec.rb
Bisect started using options: "spec/services/event_generator/geoblacklight_event_generator_spec.rb spec/services/event_generator/orangelight_event_generator_spec.rb"
Running suite to find failures...
Bisect failed!

No failures found. Bisect only works in the presence of one or more failing examples.
```

**Comment by jrgriffiniii** [comment]
https://circleci.com/gh/pulibrary/figgy/25256?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link passed without problems. I am still unable to reproduce these failures locally.

**Comment by jrgriffiniii** [comment]
Now that the test suites are passing consistently, this should truly be ready for review and testing.

File: Gemfile [hunk 1]
```diff
@@ -82,7 +82,7 @@ gem "rsolr", ">= 1.0"
 
 gem "aasm"
 gem "arabic-letter-connector"
-gem "browse-everything", github: "jrgriffiniii/browse-everything", branch: "figgy-issues-3587-jrgriffiniii"
+gem "browse-everything", github: "samvera/browse-everything", branch: "2.x-stable"
```
Comment (tpendragon) [inline Gemfile]: Is there a released version?

File: app/controllers/base_resource_controller.rb [hunk 2]
```diff
@@ -122,4 +84,39 @@ def remove_from_parent
...
+    end
+
+    def browse_everything_uploads
+      return [] unless browse_everything_params.key?("uploads")
+      @browse_everything_uploads ||= browse_everything_params["uploads"]
+    end
+
+    # Construct the pending download objects
+    # @return [Array<PendingUpload>]
+    def new_pending_uploads
```
Comment (tpendragon) [inline app/controllers/base_resource_controller.rb]: Should this be memo-ized?

File: app/controllers/bulk_ingest_controller.rb [hunk 3]
```diff
@@ -90,8 +60,124 @@ def source_metadata_id_from_path
...
+      @new_uploads ||= begin
+                         browse_everything_uploads.map do |upload_id|
+                           find_upload(upload_id)
+                         end
+                       end
+    end
+
+    def ingest_multi_volume_works(change_set_persister)
+      new_uploads.each do |upload|
+        file_tree = {}
```
Comment (tpendragon) [inline app/controllers/bulk_ingest_controller.rb]: You can make this `Hash.new([])` so you don't have to do all that || business below, just `file_tree[upload_file.container_id] += [new_pending_upload]`, I think.

File: app/controllers/bulk_ingest_controller.rb [hunk 4]
```diff
@@ -90,8 +60,124 @@ def source_metadata_id_from_path
...
+
+    def new_uploads
+      @new_uploads ||= begin
+                         browse_everything_uploads.map do |upload_id|
+                           find_upload(upload_id)
+                         end
+                       end
+    end
+
+    def ingest_multi_volume_works(change_set_persister)
```
Comment (tpendragon) [inline app/controllers/bulk_ingest_controller.rb]: This method's really big. I wonder if we can extract some classes or at least add some documentation? I'm having a hard time understanding the order of operations and what the goal is. Or maybe some refactors? Like I wonder if we can just map uploads to `PendingUpload`s and then `group_by(&:container_id)`.

File: app/controllers/bulk_ingest_controller.rb [hunk 5]
```diff
@@ -90,8 +60,124 @@ def source_metadata_id_from_path
...
+          parent_name = container.name
+
+          parent_change_set = build_change_set(title: parent_name, member_ids: member_ids)
+          change_set_persister.save(change_set: parent_change_set)
+        end
+      end
+    end
+
+    def ingest_works(change_set_persister)
+      new_uploads.each do |upload|
```
Comment (tpendragon) [inline app/controllers/bulk_ingest_controller.rb]: This seems like the same method as in `ingest_multi_volume_works`, which points again to maybe some extraction we can do. This controller's real big too.

File: app/controllers/bulk_ingest_controller.rb [hunk 6]
```diff
@@ -90,8 +60,124 @@ def source_metadata_id_from_path
...
+          children = file_tree[container.id]
+          volume_name = container.name
+          member_change_set = build_change_set(title: volume_name, pending_uploads: children, files: children)
+
+          change_set_persister.save(change_set: member_change_set)
+        end
+      end
+    end
+
+    def first_upload
```
Comment (tpendragon) [inline app/controllers/bulk_ingest_controller.rb]: Is the goal here law of demeter? I sort of wonder if the abstraction is worth it or if we should just see `new_uploads.first`

File: app/controllers/bulk_ingest_controller.rb [hunk 7]
```diff
@@ -1,7 +1,5 @@
 # frozen_string_literal: true
 class BulkIngestController < ApplicationController
```
Comment (tpendragon) [inline app/controllers/bulk_ingest_controller.rb]: A lot of things are called `uploads`. There's new uploads and pending uploads and files and containers, I wonder if you could explain the differences?

File: app/services/file_appender.rb [hunk 8]
```diff
@@ -202,5 +202,18 @@ def adjust_pending_uploads(resource)
       return unless resource.respond_to?(:pending_uploads)
       resource.pending_uploads = [] if resource.pending_uploads.blank?
       resource.pending_uploads = (resource.pending_uploads || []) - files
+
+      # This is needed for newly-created PendingUploads for uploads to the same UploadFile
```
Comment (tpendragon) [inline app/services/file_appender.rb]: I'm confused. What's an `UploadFile` and why am I uploading into it?

File: app/views/layouts/application.html.erb [hunk 9]
```diff
@@ -39,6 +39,7 @@
   <body>
     <%= content_for?(:body) ? yield(:body) : yield %>
     <%= render 'shared/ajax_modal' %>
+    <script src="//cdn.jsdelivr.net/npm/clipboard@2.0.6/dist/clipboard.js"></script>
```
Comment (tpendragon) [inline app/views/layouts/application.html.erb]: We weren't able to package this in?

File: config/application.rb [hunk 10]
```diff
@@ -10,6 +10,7 @@
 require "action_view/railtie"
 require "action_cable/engine"
 require "sprockets/railtie"
+require "active_storage"
```
Comment (tpendragon) [inline config/application.rb]: What are we using ActiveStorage for now?

File: config/application.rb [hunk 11]
```diff
@@ -29,6 +30,10 @@ class Application < Rails::Application
       allow do
         origins "*"
         resource "/graphql", headers: :any, methods: [:post]
+
+        if Rails.env.development?
+          resource "/browse/*", headers: :any, methods: [:options, :get, :post, :patch]
```
Comment (tpendragon) [inline config/application.rb]: If this is just in development do we have to add CORS for this path to nginx?

File: config/storage_services.yml [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+  service: Disk
+  root: <%= Rails.root.join("tmp/storage") %>
+
+local:
+  service: Disk
+  root: <%= Rails.root.join("storage") %>
+
+# Use rails secrets:edit to set the AWS secrets (as shared:aws:access_key_id|secret_access_key)
+# If these are enabled, ERB fails compile
+amazon:
```
Comment (tpendragon) [inline config/storage_services.yml]: We should get rid of the settings here we don't use

File: Gemfile (hunk 1 above)
Comment (hackmastera) [reply Gemfile:85]: Not yet. Making an issue to move to released branch once it's available: https://github.com/pulibrary/figgy/issues/3805

**Comment by hackmastera** [comment]
Error in development when selecting "file system":
https://gist.github.com/hackmastera/c6ffab2228adce9cf6b2d594b30bd2b9

**Comment by hackmastera** [comment]
JS console Errors in staging when submitting upload form:

```
Failed to load resource: the server responded with a status of 405 ()
scanned_resource:1 Uncaught (in promise) SyntaxError: Unexpected token < in JSON at position 0
```

**Comment by hackmastera** [comment]
TODO: 

upstream:

* namespace the classes in https://github.com/samvera/browse-everything/blob/2.x-stable/lib/browse_everything/authorization.rb#L10-L18
* Remove `mint_id` from browse everything upload creation in the browse everything controller

locally:
* Upgrade to rails 5.2

File: config/application.rb (hunk 10 above)
Comment (hackmastera) [reply config/application.rb:14]: It's a file getter

File: app/services/file_appender.rb (hunk 8 above)
Comment (hackmastera) [reply app/services/file_appender.rb:207]: we think we can get rid of this block now, due to improved equality logic in valkyrie

File: app/views/layouts/application.html.erb (hunk 9 above)
Comment (hackmastera) [reply app/views/layouts/application.html.erb:42]: Add a comment: There were bugs when we tried to include this webpack. We can add a link to the comment chain in this PR.

File: config/application.rb (hunk 11 above)
Comment (hackmastera) [reply config/application.rb:36]: Add comment: This makes development work when you're running the react front-end on a different port from the rails server. Permit the front-end to access the browse everything controllers.

File: config/storage_services.yml [hunk 13]
```diff
@@ -0,0 +1,28 @@
+test:
```
Comment (hackmastera) [inline config/storage_services.yml:1]: Add comment: Active storage configuration

File: config/storage_services.yml [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+  service: Disk
+  root: <%= Rails.root.join("tmp/storage") %>
+
+local:
+  service: Disk
+  root: <%= Rails.root.join("storage") %>
```
Comment (hackmastera) [inline config/storage_services.yml:7]: Make this an ENV so we can use a shared directory -- workers need to access these files to copy them to figgy storage.

**Comment by hackmastera** [comment]
During QA we tried the Lapidus project and found that nginx killed the request before the browse everything upload controller could complete its work. It seemed like a performance issue. After profiling and resolving what we could, it's clear that the problem is creating activerecord objects for each file. On medium-size or larger ingests the current architecture won't work.

Potential options:
- Get browse everything upload job to happen in the background and then invoke figgy somehow other than a controller redirect. Not sure what this looks like exactly, implementation-wise.
- Rewrite the BE filesystem provider to just return the root directory. We don't like this option b/c it would mess up the way we upload in the file manager.
- Have 2 buttons, one for filesystem (where you provide the root directory) and one for googledrive, which uses BE.
- Drop this for now and fix bulk ingest off master. This would leave other bugs unresolved, mostly around the google drive file-picker interaction.
- Update Nginx to not time out. This leaves a confusing user interaction: A message that says "Uploading files..." that stays for a long time, and no indication that the page is still alive.

File: spec/support/hash_provider.rb [hunk 15]
```diff
@@ -0,0 +1,68 @@
+# BrowseEverything::Provider which is powered off a simple hash. Useful for
+# testing cloud functionality without hitting the cloud.
+class HashProvider
```
Comment (hackmastera) [inline spec/support/hash_provider.rb:3]: 😍

File: app/overrides/browse_everything/upload_override.rb [hunk 16]
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+
+module BrowseEverything
+  class Upload
+    def perform_job; end
+  end
+end
```
[the remark this answers is not in the document]
Comment (tpendragon) [reply app/overrides/browse_everything/upload_override.rb:7]: `include` won't work, but I can use `prepend`.

**Pull request merged**